File: tests/cpu/julia/test_creation.jl
# Creation CPU tests — aligned with C++ test_creation.cpp
# Run with:
#   LD_LIBRARY_PATH=build/backends/cpu julia tests/cpu/julia/test_creation.jl

push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "bindings", "julia"))
push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "build", "bindings", "julia"))

using Insight

passed = 0; failed = 0
function check(name, cond)
    global passed, failed
    if cond; passed += 1; else; failed += 1; println("FAIL: $name"); end
end

println("=== Creation ===")

# zeros_1d
a = Insight.zeros(Int64[5], Insight.float32)
check("zeros_1d", Insight.numel(a) == 5)

# zeros_2d
a = Insight.zeros(Int64[3, 4], Insight.float64)
check("zeros_2d numel", Insight.numel(a) == 12)
check("zeros_2d ndim", Insight.ndim(a) == 2)

# ones_1d
a = Insight.ones(Int64[4], Insight.float32)
check("ones_1d", Insight.numel(a) == 4)

# ones_2d
a = Insight.ones(Int64[2, 5], Insight.float64)
check("ones_2d", Insight.numel(a) == 10)

# full
a = Insight.full(Int64[3, 3], 7.0, Insight.float32)
check("full", Insight.numel(a) == 9)

# full_negative
a = Insight.full(Int64[2, 2], -3.5, Insight.float64)
check("full_negative", Insight.numel(a) == 4)

# eye
a = Insight.eye(Int64(4), Int64(4), Insight.float32)
check("eye", Insight.numel(a) == 16)

# arange
a = Insight.arange(0.0, 10.0, 1.0, Insight.float64)
check("arange", Insight.numel(a) == 10)

# linspace
a = Insight.linspace(0.0, 1.0, Int64(11), Insight.float64)
check("linspace", Insight.numel(a) == 11)

# from_data
a = Insight.from_data([1.5, 2.5, 3.5])
check("from_data", Insight.numel(a) == 3)

# from_data_2d
a = Insight.from_data(Float32[1 2 3; 4 5 6])
check("from_data_2d numel", Insight.numel(a) == 6)
check("from_data_2d ndim", Insight.ndim(a) == 2)

# dtype_float64
a = Insight.zeros(Int64[3], Insight.float64)
check("dtype_float64", Insight.dtype(a) == Insight.float64)

# dtype_int32
a = Insight.ones(Int64[3], Insight.int32)
check("dtype_int32", Insight.dtype(a) == Insight.int32)

println("\n" * "="^40)
println("Results: $passed passed, $failed failed")
if failed > 0; exit(1); end

File: tests/cpu/lua/test_creation.lua
-- Creation CPU binding tests — aligned with C++ test_creation.cpp
-- Run with:
--   LUA_PATH="bindings/lua/?/init.lua;;" LUA_CPATH="build/bindings/lua/?.so;;"
--   LD_LIBRARY_PATH=build/backends/cpu ~/.luarocks/bin/busted tests/cpu/lua/test_creation.lua

describe("Creation CPU Tests", function()
  local ins
  setup(function()
    ins = require("_insight")
    ins.set_device(ins.CPUPlace())
  end)

  it("zeros_1d", function()
    local a = ins.zeros({ 5 }, ins.float32)
    assert.is_not_nil(a)
    assert.are.equal(5, a.numel)
  end)

  it("zeros_2d", function()
    local a = ins.zeros({ 3, 4 }, ins.float64)
    assert.is_not_nil(a)
    assert.are.equal(12, a.numel)
    assert.are.equal(2, a.ndim)
  end)

  it("ones_1d", function()
    local a = ins.ones({ 4 }, ins.float32)
    assert.is_not_nil(a)
    assert.are.equal(4, a.numel)
  end)

  it("ones_2d", function()
    local a = ins.ones({ 2, 5 }, ins.float64)
    assert.is_not_nil(a)
    assert.are.equal(10, a.numel)
  end)

  it("full", function()
    local a = ins.full({ 3, 3 }, 7.0, ins.float32)
    assert.is_not_nil(a)
    assert.are.equal(9, a.numel)
  end)

  it("full_negative", function()
    local a = ins.full({ 2, 2 }, -3.5, ins.float64)
    assert.is_not_nil(a)
    assert.are.equal(4, a.numel)
  end)

  it("eye", function()
    local a = ins.eye(4)
    assert.is_not_nil(a)
    assert.are.equal(16, a.numel)
  end)

  it("arange", function()
    local a = ins.arange(10, ins.float32)
    assert.is_not_nil(a)
    assert.are.equal(10, a.numel)
  end)

  it("linspace", function()
    local a = ins.linspace(0.0, 1.0, 11, ins.float64)
    assert.is_not_nil(a)
    assert.are.equal(11, a.numel)
  end)

  it("from_table_1d", function()
    local a = ins.from_table({ 1.5, 2.5, 3.5 })
    assert.is_not_nil(a)
    assert.are.equal(3, a.numel)
  end)

  it("from_table_2d", function()
    local a = ins.from_table({ { 1, 2, 3 }, { 4, 5, 6 } })
    assert.is_not_nil(a)
    assert.are.equal(6, a.numel)
    assert.are.equal(2, a.ndim)
  end)

  it("zeros_like", function()
    local a = ins.ones({ 3, 3 }, ins.float32)
    local b = ins.zeros_like(a)
    assert.is_not_nil(b)
    assert.are.equal(9, b.numel)
  end)

  it("ones_like", function()
    local a = ins.zeros({ 2, 4 }, ins.float64)
    local b = ins.ones_like(a)
    assert.is_not_nil(b)
    assert.are.equal(8, b.numel)
  end)

  it("dtype_float64", function()
    local a = ins.zeros({ 3 }, ins.float64)
    assert.are.equal(ins.float64, a.dtype)
  end)

  it("dtype_int32", function()
    local a = ins.ones({ 3 }, ins.int32)
    assert.are.equal(ins.int32, a.dtype)
  end)

  -- ====================================================================
  -- Error handling: must throw Lua errors, NOT crash the process
  -- ====================================================================

  it("error_ragged_table", function()
    assert.has_error(function()
      ins.Array({ { 3, 3 }, { 3 } })
    end)
  end)

  it("error_ragged_deep", function()
    assert.has_error(function()
      ins.Array({ { { 1, 2 }, { 3, 4 } }, { { 5, 6 } } })
    end)
  end)

  it("error_string_element", function()
    assert.has_error(function()
      ins.Array({ 1, 2, "hello" })
    end)
  end)

  it("error_nil_element", function()
    assert.has_error(function()
      ins.Array({ 1, nil, 3 })
    end)
  end)

  it("error_from_table_ragged", function()
    assert.has_error(function()
      ins.from_table({ { 1, 2 }, { 3 } })
    end)
  end)

  it("valid_flat_table", function()
    local a = ins.Array({ 1, 2, 3, 4 })
    assert.is_not_nil(a)
    assert.are.equal(4, a.numel)
  end)

  it("valid_nested_table", function()
    local a = ins.Array({ { 1, 2 }, { 3, 4 } })
    assert.is_not_nil(a)
    assert.are.equal(4, a.numel)
    assert.are.equal(2, a.ndim)
  end)

  it("process_survives_ragged_error", function()
    -- Attempt a ragged table, then verify the process is still alive
    pcall(function()
      ins.Array({ { 1, 2 }, { 3 } })
    end)
    local a = ins.Array({ 5, 6, 7 })
    assert.is_not_nil(a)
    assert.are.equal(3, a.numel)
  end)
end)
